### PR TITLE
Only apply one edit when rolling back buffer before commit

### PR DESCRIPTION
**Pull request by rchande**
In https://devdiv.visualstudio.com/DevDiv/_workitems?id=242397&_a=edit we spend additional time in WpfTextView reacting to subjectbuffer changes while rolling back the text buffer to prepare to insert a completion item. We currently apply an edit for each snapshot version between the trigger and current snapshots. We can instead apply all the buffer diffs in one edit, which may reduce time in WpfTextView.OnVisualBufferChanged.

This is all under one undo transaction (http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Intellisense/Completion/Controller_Commit.cs,55) so the undo behavior should stay the same.

Tag @dotnet/roslyn-ide @dotnet/roslyn-perf for review

```
Name
 microsoft.codeanalysis.editorfeatures!Controller.RollbackToBeforeTypeChar
+ microsoft.visualstudio.platform.vseditor.ni!Microsoft.VisualStudio.Text.Implementation.BaseBuffer+TextBufferEdit.Apply()
|+ microsoft.visualstudio.platform.vseditor.ni!BufferGroup.FinishEdit
||+ microsoft.visualstudio.platform.vseditor.ni!Microsoft.VisualStudio.Text.Implementation.BaseBuffer+TextContentChangedEventRaiser.RaiseEvent(Microsoft.VisualStudio.Text.Implementation.BaseBuffer, Boolean)
|||+ microsoft.visualstudio.platform.vseditor.ni!BaseBuffer.RawRaiseEvent
||| + microsoft.visualstudio.platform.vseditor.ni!Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[System.__Canon](System.Object, System.EventHandler`1, System.__Canon)
|||  + microsoft.visualstudio.platform.vseditor.ni!WpfTextView.OnVisualBufferChanged
```

### CODE REVIEW

**Comment by rchande** [comment]
test vsi

**Comment by rchande** [comment]
test vsi please

**Comment by rchande** [comment]
Note that VSI tests are failing due to a build issue. Known ( @tmeschter )?

**Comment by rchande** [comment]
@CyrusNajmabadi Can you think of any cases where this will be a problem?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 1]
```diff
@@ -150,20 +150,21 @@ private void RollbackToBeforeTypeChar(ITextSnapshot initialTextSnapshot)
...
+            using (var textEdit = this.SubjectBuffer.CreateEdit(EditOptions.None, reiteratedVersionNumber: null, editTag: null))
             {
-                var version = versions[i];
-                using (var textEdit = this.SubjectBuffer.CreateEdit(EditOptions.None, reiteratedVersionNumber: null, editTag: null))
+                // Un-apply the edits. 
+                for (var i = versions.Count - 1; i >= 0; i--)
                 {
-                    foreach (var change in version.Changes)
+                    var version = versions[i];
                     {
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: this block looks unnecssary.

**Comment by CyrusNajmabadi** [comment]
So, this is @jasonmalinowski 's area of expertise.  I don't have an issue with this per-se... but i'm certainly not filled with confidence that this is particularly safe.

**Comment by jasonmalinowski** [comment]
I wouldn't expect this to work. When you start an text buffer edit, and apply it, all the operations you do are relative to the _original_ snapshot. So the series of operations of:

1. Create a new edit.
2. Insert at position 0 "asdf"
3. Insert at position 10, "jkl;"
4. Complete the edit.

Means that the 'j' being inserted here isn't going to be at position 10, but rather position 4, since it was "pushed" by the first insertion at the same time. This is why we were applying the edits as we went. If we are trying to undo revisions 3, 4, 5 and 6, you can't undo 5 until you have the buffer's position after reverting 6, etc, etc. I suspect your code will happen to work in a type-forward situation, but do something like:

1. Type out Console.rite (verbatim)
2. Invoke completion on 'rite'
3. While the session is still active, arrow over to the end and type Lin (as if you're completing to WriteLine)
4. While this session is active, arrow to the start type a W
5. Arrow back to the end, commit with a (.
6. Undo this.

...and I expect this to do something hilarious. Previously, we'd be reverting the insertion of "Lin" in the coordinate system relative to after the W was typed. Now we're reverting it simultaneously.

The desire to merge many edits back into one is a famous problem. The fully general case where you want optimal outputs is really hard. It took Jack weeks to recover when he tried. It looks like we have the original snapshot though, and in our constrained scenario we might be able to get lucky. You can loop through the edits (forward is probably easier) and build up a single before/after span that encompasses all the edits. Then just do `edit.Replace(newSpan, originalSnapshot.GetText(originalSpan))`, assuming we don't need to do some tracking of the caret in the middle of that.

**Comment by Pilchie** [comment]
Also - write a test that captures the break that @jasonmalinowski describes above :)

@jmarolf I am suspicious that this undo invoking code is leading to your text buffer edit in progress bugs too, since it's relatively new to the completion code.

**Comment by rchande** [comment]
@jasonmalinowski The particular scenario you described doesn't do anything unusual but I will keep investigating
undo

**Comment by rchande** [comment]
@jasonmalinowski Integration tests added by https://github.com/dotnet/roslyn-internal/issues/1504 pass with this change. Any other feedback?

**Comment by jasonmalinowski** [comment]
Oh, I see. This doesn't roll back to the _original_ triggering snapshot, it juts rolls back to before the typechar. In the cases I was dreaming up, it was already only rolling back one version so the bug isn't noticeable. Do we actually have scenarios where that happens? The only one I could find with the debugger is brace completion, where we just happen to stick the second edit right after the first, so this works by pure dumb luck. My theory therefore is:

1. In the case of a non-trivial edit, this is still broken.
2. Those scenarios don't exist anyways, so this isn't helping perf like you think it is.

So yes, this is still broken in the face of extensions, and I'm also not convinced it helps outside the brace completion scenario, and then only it's still by luck. My status stays the same until we at least do a check that this _is_ the safe special case of brace completion, and you also prove that brace completion was being used enough in your perf scenario that this is actually going to help.

**Comment by rchande** [comment]
This is probably not the right fix. Closing.

**Pull request closed without merging**